<launch>
  <!-- 实验：仅启动相机 + ArUco + 多机器人融合，可选打开 RViz/调试节点 -->

  <arg name="robots_config" default="$(find jaka_close_contro)/config/robots.yaml" />
  <arg name="robot_ids" default="$(eval [r['id'] for r in load_yaml(arg('robots_config'))['robots']])" />
  <arg name="faces_yaml" default="$(find jaka_close_contro)/config/cube_faces_4robots.yaml" />
  <arg name="svo_file" default="" />
  <arg name="use_debug_node" default="true" />

  <arg name="enable_jaka1" default="$(eval '1' in str(arg('robot_ids')).replace('[',' ').replace(']',' ').replace(',',' ').split())" />
  <arg name="enable_jaka2" default="$(eval '2' in str(arg('robot_ids')).replace('[',' ').replace(']',' ').replace(',',' ').split())" />
  <arg name="enable_jaka3" default="$(eval '3' in str(arg('robot_ids')).replace('[',' ').replace(']',' ').replace(',',' ').split())" />
  <arg name="enable_jaka4" default="$(eval '4' in str(arg('robot_ids')).replace('[',' ').replace(']',' ').replace(',',' ').split())" />

  <!-- 1. ZED2i 相机 -->
  <include file="$(find zed_wrapper)/launch/zed2i.launch">
    <arg name="svo_file" value="$(arg svo_file)" />
  </include>

  <!-- 2. ArUco 检测与分流 -->
  <node pkg="jaka_close_contro" type="cube_aruco_detector_node" name="cube_aruco_detector" output="screen">
    <param name="image_topic"       value="/zed2i/zed_node/left/image_rect_color" />
    <param name="camera_info_topic" value="/zed2i/zed_node/left/camera_info" />
    <param name="fiducial_topic"    value="/fiducial_transforms" />
    <param name="default_marker_size" value="0.08" />
    <param name="dictionary" value="DICT_4X4_50" />
    <param name="tool_dictionary" value="DICT_6X6_1000" />
    <param name="tool_marker_length_m" value="0.0425" />
    <param name="tool_marker_separation_m" value="0.005" />
    <param name="min_gridboard_markers" value="2" />
    <param name="world_board_enable" value="true" />
    <param name="world_board_output_id" value="500" />
    <param name="world_board_marker_length_m" value="0.0525" />
    <param name="world_board_marker_separation_m" value="0.005" />
    <param name="world_board_min_markers" value="2" />
    <param name="enable_single_world_markers" value="false" />
    <param name="robot_stride" value="100" />
    <param name="id_step" value="10" />
    <param name="face_id_base" value="10" />
    <rosparam subst_value="true" param="robot_ids">$(arg robot_ids)</rosparam>
    <rosparam param="marker_sizes">
      "0": 0.08
    </rosparam>
    <rosparam param="world_board_marker_ids">[500, 501, 502, 503]</rosparam>
  </node>

  <node pkg="jaka_close_contro" type="fiducial_relay_node" name="fiducial_relay" output="screen">
    <rosparam param="world_ids">[500, 501, 502, 503]</rosparam>
    <rosparam subst_value="true" param="robot_ids">$(arg robot_ids)</rosparam>
    <param name="input_topic" value="/fiducial_transforms" />
    <param name="output_world_topic" value="/world_fiducials" />
    <param name="output_tool_topic" value="/tool_fiducials" />
  </node>

  <!-- 3. 世界坐标系定义 -->
  <node pkg="jaka_close_contro" type="world_tag_node" name="world_tag_node" output="screen">
    <param name="world_id" type="int" value="500" />
    <rosparam param="world_ids">[500, 501, 502, 503]</rosparam>
    <param name="fiducial_topic"    value="/world_fiducials" />
    <param name="world_frame"       value="world" />
    <param name="camera_frame"      value="zed2i_left_camera_optical_frame" />
    <param name="alpha_pos"         value="0.3" />
    <param name="alpha_rot"         value="0.3" />
  </node>

  <!-- 4. 多面融合 -->
  <group if="$(arg enable_jaka1)">
    <node pkg="jaka_close_contro" type="cube_multi_face_fusion_node" name="cube_multi_face_fusion_jaka1" output="screen">
      <param name="faces_yaml"           value="$(arg faces_yaml)" />
      <param name="fiducial_topic"       value="/tool_fiducials" />
      <param name="cube_frame"           value="cube_center_jaka1" />
      <param name="camera_frame_default" value="zed2i_left_camera_optical_frame" />
      <param name="timeout_sec"          value="0.5" />
      <param name="robot_id" value="1" />
      <param name="robot_name" value="jaka1" />
      <param name="face_id_base" value="10" />
      <param name="robot_stride" value="100" />
      <param name="output_topic" value="/vision/jaka1/cube_center" />
      <param name="stats_topic" value="/vision/jaka1/cube_fusion_stats" />
      <param name="filter_robot_faces" value="true" />
      <param name="broadcast_tf" value="true" />
    </node>
    <node if="$(arg use_debug_node)" pkg="jaka_close_contro" type="cube_fusion_debug_node" name="cube_fusion_debug_jaka1" output="screen">
      <param name="faces_yaml"      value="$(arg faces_yaml)" />
      <param name="fiducial_topic"  value="/tool_fiducials" />
      <param name="fused_pose_topic" value="/vision/jaka1/cube_center" />
      <param name="camera_frame" value="zed2i_left_camera_optical_frame" />
      <param name="fused_frame" value="cube_center_jaka1" />
      <param name="publish_predicted_pose" value="true" />
    </node>
  </group>

  <group if="$(arg enable_jaka2)">
    <node pkg="jaka_close_contro" type="cube_multi_face_fusion_node" name="cube_multi_face_fusion_jaka2" output="screen">
      <param name="faces_yaml"           value="$(arg faces_yaml)" />
      <param name="fiducial_topic"       value="/tool_fiducials" />
      <param name="cube_frame"           value="cube_center_jaka2" />
      <param name="camera_frame_default" value="zed2i_left_camera_optical_frame" />
      <param name="timeout_sec"          value="0.5" />
      <param name="robot_id" value="2" />
      <param name="robot_name" value="jaka2" />
      <param name="face_id_base" value="10" />
      <param name="robot_stride" value="100" />
      <param name="output_topic" value="/vision/jaka2/cube_center" />
      <param name="stats_topic" value="/vision/jaka2/cube_fusion_stats" />
      <param name="filter_robot_faces" value="true" />
      <param name="broadcast_tf" value="true" />
    </node>
    <node if="$(arg use_debug_node)" pkg="jaka_close_contro" type="cube_fusion_debug_node" name="cube_fusion_debug_jaka2" output="screen">
      <param name="faces_yaml"      value="$(arg faces_yaml)" />
      <param name="fiducial_topic"  value="/tool_fiducials" />
      <param name="fused_pose_topic" value="/vision/jaka2/cube_center" />
      <param name="camera_frame" value="zed2i_left_camera_optical_frame" />
      <param name="fused_frame" value="cube_center_jaka2" />
      <param name="publish_predicted_pose" value="true" />
    </node>
  </group>

  <group if="$(arg enable_jaka3)">
    <node pkg="jaka_close_contro" type="cube_multi_face_fusion_node" name="cube_multi_face_fusion_jaka3" output="screen">
      <param name="faces_yaml"           value="$(arg faces_yaml)" />
      <param name="fiducial_topic"       value="/tool_fiducials" />
      <param name="cube_frame"           value="cube_center_jaka3" />
      <param name="camera_frame_default" value="zed2i_left_camera_optical_frame" />
      <param name="timeout_sec"          value="0.5" />
      <param name="robot_id" value="3" />
      <param name="robot_name" value="jaka3" />
      <param name="face_id_base" value="10" />
      <param name="robot_stride" value="100" />
      <param name="output_topic" value="/vision/jaka3/cube_center" />
      <param name="stats_topic" value="/vision/jaka3/cube_fusion_stats" />
      <param name="filter_robot_faces" value="true" />
      <param name="broadcast_tf" value="true" />
    </node>
    <node if="$(arg use_debug_node)" pkg="jaka_close_contro" type="cube_fusion_debug_node" name="cube_fusion_debug_jaka3" output="screen">
      <param name="faces_yaml"      value="$(arg faces_yaml)" />
      <param name="fiducial_topic"  value="/tool_fiducials" />
      <param name="fused_pose_topic" value="/vision/jaka3/cube_center" />
      <param name="camera_frame" value="zed2i_left_camera_optical_frame" />
      <param name="fused_frame" value="cube_center_jaka3" />
      <param name="publish_predicted_pose" value="true" />
    </node>
  </group>

  <group if="$(arg enable_jaka4)">
    <node pkg="jaka_close_contro" type="cube_multi_face_fusion_node" name="cube_multi_face_fusion_jaka4" output="screen">
      <param name="faces_yaml"           value="$(arg faces_yaml)" />
      <param name="fiducial_topic"       value="/tool_fiducials" />
      <param name="cube_frame"           value="cube_center_jaka4" />
      <param name="camera_frame_default" value="zed2i_left_camera_optical_frame" />
      <param name="timeout_sec"          value="0.5" />
      <param name="robot_id" value="4" />
      <param name="robot_name" value="jaka4" />
      <param name="face_id_base" value="10" />
      <param name="robot_stride" value="100" />
      <param name="output_topic" value="/vision/jaka4/cube_center" />
      <param name="stats_topic" value="/vision/jaka4/cube_fusion_stats" />
      <param name="filter_robot_faces" value="true" />
      <param name="broadcast_tf" value="true" />
    </node>
    <node if="$(arg use_debug_node)" pkg="jaka_close_contro" type="cube_fusion_debug_node" name="cube_fusion_debug_jaka4" output="screen">
      <param name="faces_yaml"      value="$(arg faces_yaml)" />
      <param name="fiducial_topic"  value="/tool_fiducials" />
      <param name="fused_pose_topic" value="/vision/jaka4/cube_center" />
      <param name="camera_frame" value="zed2i_left_camera_optical_frame" />
      <param name="fused_frame" value="cube_center_jaka4" />
      <param name="publish_predicted_pose" value="true" />
    </node>
  </group>

  <!-- 5. 打开 RViz，自动加载显示配置 -->
  <node pkg="rviz" type="rviz" name="rviz" output="screen" args="-d $(find jaka_close_contro)/rviz/cube_fusion_debug.rviz" />
</launch>
